src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class BatchNode<TInput> : IIncrementalGeneratorNode<ImmutableArray<TInput>>
+    {
+        private readonly IIncrementalGeneratorNode<TInput> _sourceNode;
+        private readonly IEqualityComparer<ImmutableArray<TInput>> _comparer;
+
+        public BatchNode(IIncrementalGeneratorNode<TInput> sourceNode, IEqualityComparer<ImmutableArray<TInput>>? comparer = null)
+        {
+            _sourceNode = sourceNode;
+            _comparer = comparer ?? EqualityComparer<ImmutableArray<TInput>>.Default;
+        }
+
+        public IIncrementalGeneratorNode<ImmutableArray<TInput>> WithComparer(IEqualityComparer<ImmutableArray<TInput>> comparer) => new BatchNode<TInput>(_sourceNode, comparer);
```

Comment:
My instinct here would be to take `IEqualityComparer<TInput>`, not over the array. Was this something that came out of the API review? If not I'd like to chat about this a bit later. Don't want it to hold up the merge to main.